src/main/kotlin/org/jetbrains/research/libsl/visitors/BlockStatementVisitor.kt

```diff
@@ -0,0 +1,133 @@
+                statements.add(assignmentWithLeftUnaryOp)
+            }
+            ctx.rightUnaryOp != null -> let {
+                val expressionVisitor = ExpressionVisitor(functionContext)
+                val op = ArithmeticUnaryOp.fromString(ctx.rightUnaryOp.text)
+                val value = expressionVisitor.visitQualifiedAccess(ctx.qualifiedAccess())
+                val assignmentWithRightUnaryOp = AssignmentWithRightUnaryOp(op, value)
+                statements.add(assignmentWithRightUnaryOp)
+            }
+        }
+    }
+
+    override fun visitIfStatement(ifCtx: LibSLParser.IfStatementContext) {
+        val expressionVisitor = ExpressionVisitor(functionContext)
+        val value = expressionVisitor.visitExpression(ifCtx.expression())
+
+        val ifStatements = mutableListOf<Statement>()
+        val ifStatementVisitor = BlockStatementVisitor(functionContext, ifStatements, localVariables)
+        ifCtx.functionBodyStatements().forEach { ifStatementVisitor.visit(it) }
+
+        val elseStatement = ifCtx.elseStatement()?.let { elseStmt ->
+            val elseStatements = mutableListOf<Statement>()
+            val elseStatementsVisitor = BlockStatementVisitor(functionContext, elseStatements, localVariables)
+            elseStmt.functionBodyStatements().forEach { elseStatementsVisitor.visit(it) }
+
+            ElseStatement(elseStatements)
+        }
+
+        val ifBlock = IfStatement(value, ifStatements, elseStatement)
+
+        statements.add(ifBlock)
+    }
+
+    override fun visitAction(ctx: LibSLParser.ActionContext) {
+        val name = ctx.Identifier().text.extractIdentifier()
+        val expressionVisitor = ExpressionVisitor(functionContext)
+        val args = ctx.expressionsList().expression().map { expr ->
+            expressionVisitor.visitExpression(expr)
+        }.toMutableList()
+
+        val action = Action(name, args)
+
+        statements.add(action)
+    }
+
+    override fun visitProc(ctx: LibSLParser.ProcContext) {
+        val name = ctx.Identifier().text.extractIdentifier()
+        val expressionVisitor = ExpressionVisitor(functionContext)
+        val args = ctx.expressionsList()?.expression()?.map { expr ->
+            expressionVisitor.visitExpression(expr)
+        }?.toMutableList()
+        val hasThisExpression = ctx.THIS() != null
+        val hasParentExpression = ctx.PARENT() != null
+
+        val proc = Proc(name, args, hasThisExpression, hasParentExpression)
+
+        statements.add(proc)
+    }
+
+    override fun visitVariableDecl(ctx: LibSLParser.VariableDeclContext) {
+        val keyword = VariableKeyword.fromString(ctx.keyword.text)
+        val name = ctx.nameWithType().name.asPeriodSeparatedString()
+        val typeReference = processTypeIdentifier(ctx.nameWithType().type)
+        val expressionVisitor = ExpressionVisitor(context)
+        val initValue = ctx.assignmentRight()?.let { right ->
```

Comment:
The comment about parser refactoring for variable assignment is suitable here too